File: THCS/2022_2023/YenBai_2223/Cau3/KHIEUVU.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>

using namespace std;

int countDancingPairs(vector<int>& heights, int k) {
    sort(heights.begin(), heights.end()); // Sắp xếp danh sách chiều cao
    int count = 0;
    int n = heights.size();

    int left = 0;
    int right = 0;

    while (left < n) {
        // Tìm kiếm vị trí của right
        while (right < n && heights[right] - heights[left] < k) {
            right++;
        }

        // Nếu chênh lệch chiều cao bằng k
        if (right < n && heights[right] - heights[left] == k) {
            // Đếm số chiều cao giống nhau ở bên trái
            int countLeft = 1;
            while (left + 1 < n && heights[left] == heights[left + 1]) {
                countLeft++;
                left++;
            }

            // Đếm số chiều cao giống nhau ở bên phải
            int countRight = 1;
            while (right + 1 < n && heights[right] == heights[right + 1]) {
                countRight++;
                right++;
            }

            // Cộng vào tổng số cặp
            count += countLeft * countRight;
        }

        left++;
    }

    return count;
}

int main() {
    ifstream infile("KHIEUVU.INP");
    ofstream outfile("KHIEUVU.OUT");

    int n, k;
    infile >> n >> k;

    vector<int> heights(n);
    for (int i = 0; i < n; i++) {
        infile >> heights[i];
    }

    // Tính số cặp đôi và ghi vào tệp
    int result = countDancingPairs(heights, k);
    outfile << result << endl;

    infile.close();
    outfile.close();

    return 0;
}

File: THCS/2022_2023/YenBai_2223/Cau3/KHIEUVU.py
def count_dancing_pairs(heights, k):
    """Đếm số cặp đôi có độ chênh lệch chiều cao bằng k."""
    heights.sort()  # Sắp xếp danh sách chiều cao
    count = 0
    n = len(heights)
    
    left = 0
    right = 0
    
    while right < n:
        # Tìm kiếm các cặp có độ chênh lệch chiều cao bằng k
        while right < n and heights[right] - heights[left] < k:
            right += 1
        # Nếu chênh lệch chiều cao bằng k, đếm các cặp
        if right < n and heights[right] - heights[left] == k:
            # Đếm số chiều cao giống nhau ở bên trái
            count_left = 1
            while left + 1 < n and heights[left] == heights[left + 1]:
                count_left += 1
                left += 1
            
            # Đếm số chiều cao giống nhau ở bên phải
            count_right = 1
            while right + 1 < n and heights[right] == heights[right + 1]:
                count_right += 1
                right += 1
            
            # Cộng vào tổng số cặp
            count += count_left * count_right
        
        left += 1

    return count

# Đọc dữ liệu từ tệp
with open("KHIEUVU.INP", "r") as infile:
    n, k = map(int, infile.readline().strip().split())
    heights = list(map(int, infile.readline().strip().split()))

# Tính số cặp đôi và ghi vào tệp
result = count_dancing_pairs(heights, k)

# Ghi kết quả vào tệp
with open("KHIEUVU.OUT", "w") as outfile:
    outfile.write(f"{result}\n")
